File: src/dsLinkedList/LinkedList.java
import java.util.NoSuchElementException;

public class LinkedList{
    private class Node{
        private int value;
        private Node next;

        public Node(int value){
            this.value = value;
        }
    }
    private Node first;
    private Node last;
    private int size;
    public void addLast(int item){
        var node = new Node(item);
        if(isEmpty())
            first = last = node;
        else{
            last.next = node;
            last = node;
        }
        size++;
    }
    public void addFirst(int item){
        var node = new Node(item);
        if(isEmpty())
            first = last = node;
        else{
            node.next = first;
            first = node;
        }
        size++;
    }
    public int indexOf(int item){
        int index =0;
        var current = first;
        while(current!=null){
            if(current.value == item)
                return index;
            current = current.next;
            index++;
        }
        return -1;
    }
    public void removeFirst(){
        if(isEmpty())
            throw new NoSuchElementException();
        if(first==last) {
            first = last = null;
            return;
        }
        else{
            var second = first.next;
            first.next = null;
            first = second;
        }
        size--;
    }
    public void removeLast(){
        if(isEmpty())
            throw new NoSuchElementException();
        if(first == last){
            first = last = null;
            return;
        }
        else{
            var previous = getPrevious(last);
            last = previous;
            last.next = null;
        }
        size--;
    }
    public void remove(int item){
        if(isEmpty())
            throw new NoSuchElementException();
        if(first.value == item){
            removeFirst();
            return;
        }
        if(last.value == item){
            removeLast();
            return;
        }
        var current = first.next;
        while (current!=null) {
            if(current.value == item){
                getPrevious(current).next = current.next;
                current = null;
                size--;
                break;
            }
            current = current.next;
        }
    }
    public boolean contains(int item){
        return indexOf(item)!= -1;
    }
    public int size(){
        return size;
    }
    public int[] toArray(){
        int[] array = new int[size];
        var current = first;
        var index = 0;
        while(current != null){
            array[index++] = current.value;
            current = current.next;
        }
        return array;
    }
    public void reverse(){
        if(isEmpty()) return;
        if(first == last) return;
        var previous = first;
        var current = first.next;
        while(current != null){
            var next = current.next;
            current.next = previous;
            previous = current;
            current = next;
        }
        last = first;
        last.next = null;
        first = previous;
    }
    public int getKthFromTheEnd(int k){
        if(isEmpty())
            throw new IllegalStateException();
        if(k<0 || k==0){
            System.err.println("Invalid value of k!\nk must be greater than 0 and less than size of the linkedlist.");
            throw new IllegalArgumentException();
        }
        var previous = first;
        var next = first;
        for(int i=0; i<k; i++) {
            next = next.next;
            if(next==null){
                System.err.println("Invalid value of k!\nk must be greater than 0 and less than size of the linkedlist.");
                throw new IllegalArgumentException();
            }
        }
        while(next!=null){
            previous = previous.next;
            next = next.next;
        }
        return previous.value;
    }
    private boolean isEmpty(){
        return first==null;
    }
    private Node getPrevious(Node node){
        var current = first;
        while(current != null){
            if(current.next == node) return current;
            current = current.next;
        }
        return null;
    }
}

File: src/dsLinkedList/LinkedList.cpp
#include <iostream>
#include <bits/c++io.h>
using namespace std;

class Node
{
public:
    int data;
    Node *next;
    Node *previous;

    Node()
    {
        data = 0;
        next = NULL;
    }

    Node(int item)
    {
        this->data = item;
        this->next = NULL;
    }
};

class LinkedList
{
    Node *first;
    int size = 0;

public:
    LinkedList()
    {
        first = NULL;
    }
    void insertAtLast(int item);    // Done
    void insertAtFirst(int item);   // Done
    int sizeOfList();               // Done
    void display();                 // Done
    void displayArray(int *array);  // Done
    bool exists(int item);          // Done
    int valueAtIndex(int index);    // Done
    int getPreviousValue(int item); // Done
    int *toArray();                 // Done
    Node getPrevious(Node *node);
    void removeFromLast();
    void removeFromFirst();
};

void LinkedList::insertAtLast(int item)
{
    Node *last = first;
    Node *node = new Node(item);
    if (first == NULL)
    {
        first = last = node;
        size++;
        return;
    }
    while (last->next != NULL)
    {
        last = last->next;
    }
    last->next = node;
    size++;
}

void LinkedList::insertAtFirst(int item)
{
    Node *last = first;
    Node *node = new Node(item);
    if (first == NULL)
    {
        first = last = node;
        size++;
        return;
    }
    node->next = first;
    first = node;
    size++;
}

int LinkedList::sizeOfList()
{
    return size;
}

void LinkedList::display()
{
    Node *last = first;
    while (last != NULL)
    {
        cout << last->data << " -> ";
        last = last->next;
    }
    cout << "NULL" << endl;
}

void LinkedList::displayArray(int *array)
{
    cout << "[";
    for (int i = 0; i < size - 1; i++)
    {
        cout << array[i] << ", ";
    }
    cout << array[size - 1] << "]" << endl;
}

bool LinkedList::exists(int item)
{
    Node *last = first;
    if (first == NULL)
    {
        cout << "List is empty" << endl;
        return false;
    }
    while (last != NULL)
    {
        if (last->data == item)
        {
            return true;
        }
        last = last->next;
    }
    return false;
}

int LinkedList::valueAtIndex(int index)
{
    if (index < 0 || index >= size)
    {
        cout << "Invalid index" << endl;
        return -1;
    }
    Node *last = first;
    for (int i = 0; i < index; i++)
    {
        last = last->next;
    }
    return last->data;
}

int LinkedList::getPreviousValue(int item)
{
    Node *last = first;
    while (last->next->data != item)
    {
        last = last->next;
    }
    return last->data;
}

Node LinkedList::getPrevious(Node *node)
{
    Node 
}

int *LinkedList::toArray()
{
    int *ptr = new int[size];
    Node *last = first;
    int i = 0;
    while (last != NULL)
    {
        ptr[i] = last->data;
        i++;
        last = last->next;
    }
    return ptr;
}

int main()
{
    LinkedList newLinkedList;
    // newLinkedList.insertAtLast(10);
    newLinkedList.insertAtLast(20);
    newLinkedList.insertAtLast(30);
    newLinkedList.insertAtLast(40);
    newLinkedList.insertAtLast(50);
    newLinkedList.insertAtLast(60);
    newLinkedList.insertAtLast(70);
    newLinkedList.insertAtFirst(10);
    // newLinkedList.insertAtFirst(10);
    newLinkedList.display();
    int *array = newLinkedList.toArray();
    newLinkedList.displayArray(array);

    return 0;
}
